### DISCUSSION TITLE
Set job metadata in worker process rather than work-horse

### DISCUSSION THREAD

**Discussion by y4n9squared** [rq/rq#1705] [Ideas]
The worker's `prepare_job_execution` function is where a lot of metadata about a job is set (e.g. name of worker currently processing job). If the work-horse process terminates unexpectedly, e.g. due to a segfault or OOM, the saving of the metadata does not occur since the work-horse's `finally` clause does not get to run and data is not committed back to Redis.

This can be reproduced by running a job that calls `os._exit(1)`.  It would be nice if job metadata were not lost to aid in debugging. Two ways I can think of doing this are:

1. Commit the job metadata before the job workload is run. This would still miss updates that are performed after the job completes.
2. Move job metadata updates to the worker process, leaving only deserialization of the function and args to the work-horse.

**Comment by y4n9squared** [comment]
@selwin Any thoughts on this?

**Comment by selwin** [comment]
I think approach 1 is better. We can update the metadata in a pipeline along with heartbeats etc.
